### PR TITLE
Unpooling Layer

### PR DESCRIPTION
**Pull request by jeffdonahue**
This adds my implementation of `UnpoolingLayer` which (approximately) reverses the operation of `PoolingLayer`.   I wrote this a while ago and haven't looked at any of the other implementations.

The `UnpoolingLayer` takes 1-3 bottoms.  The first (required) for both pooling types is the inputs to be unpooled (e.g. the pooled outputs from a `PoolingLayer`).

For max unpooling, there is a second required input -- the pooling mask (an optional output of `PoolingLayer` -- the second one).  Finally, there is a third input which is only required for *overlapping* max (un)pooling -- in the code I called this the "argmax count".  This input has the shape of the Unpooling *output* and the corresponding Pooling *input*, and indicates for each pooling input the number of times that input was the max.  (This PR also modifies PoolingLayer to output the argmax count as an optional 3rd top, like the mask is currently an optional 2nd top.) This quantity is always 0 or 1 for non-overlapping pooling, but may be >1 for some inputs for overlapping pooling.  For non-overlapping pooling, this input is optional and will only be used to shape the output if specified.

For average unpooling, an extra (second) input blob is also optional and if specified is used only to shape the output.

### CODE REVIEW

**Comment by bluechill52** [comment]
Hi, why is it still not added to the caffe master branch ? Can I use the layer anyway ? Thanks.

**Comment by suruoxi** [comment]
FYI

HyeonwooNoh implemented a unpooling layer.  The code is here.
https://github.com/HyeonwooNoh/caffe/blob/master/src/caffe/layers/unpooling_layer.cpp

This code was used in his paper "Learning Deconvolution Network for Semantic Segmentation"  http://cvlab.postech.ac.kr/research/deconvnet/

**Comment by VladTur** [comment]
Hi Jeff, thank you so much for this implementation!
Would you like to be so kind to tell how can I use correctly the unpooling layer for the following example:
I have 8x8 image and by pooling (MAX, kernel:2, stride:2) I got pooled image 4x4. So my pooling mask is 4x4 too. How to run unpooling (MAX, what kernel, stride, pad?) to correctly restore 8x8 image? 
When I run (MAX, kernel:1, stride:1, pad:0) it restores 4x4 which is basically not correct
When I run (MAX, kernel:2, stride:2, pad:0) there is an error:
Check failed: expected_pooled_height == pooled_height_ (2 vs. 4)
Also how to run pooling-unpooling pair correctly in a case of AVE-pooling operation for the above example?
Thank you in advance!

**Comment by twtygqyy** [comment]
Hi @VladTur 
Did you find how to make the unpooling layer work? I've tried it but it seems that the pooled height and weight keep unchanged.

**Comment by kaaicheung** [comment]
I got the same error!! Check failed: expected_pooled_height == pooled_height_
Anyone figured out how to use this unpooling layer?

**Comment by RupeshD** [comment]
Same error. Did you guys figure out?

**Comment by vgenty** [comment]
@RupeshD I was unsuccessful with Jeff's implementation, same error as the others. It's not as simple as ignoring `CHECK_EQ` expected vs. pooling height and width. I fiddled with the layer a bit and was still unsuccessful. HyeonwooNoh's unpooling layer worked fine after inserting into my own caffe fork.  It does what it says it does if you dump the unpooled tensors and visually inspect them.

**Comment by RupeshD** [comment]
@vgenty Thanks for your reply. I used below branch of caffe. It has HyeonwooNoh's unpooling layer implementation. 
https://github.com/matthieudelaro/caffeBVLCplus

**Comment by ryanflower** [comment]
For those who couldn't work with this implementation due to shape check failed, I suggest you read the source code and the corresponding unit test for the correct usage of the layer. When doing max unpooling, there should be 3 bottom blobs, and the last one is purely for getting the expected shape of the unspooling result.

The implementation is fully functional and I hope this commit to be merged to master soon.

**Comment by ChiZhangRIT** [comment]
@ryanflower I tried to add the third bottom layer as "argmax_count" in unpooling layer but I still got the error "Check failed: expected_pooled_height == pooled_height_ " and the python kernel was dead when the network goes to unpooling layer. Do you have an idea what the problem is ?

`# pool5
layer { 
  name: "pool5"		type: "Pooling"		bottom: "conv5_3"	top: "pool5"
									top: "pool5_mask"
									top: "pool5_argmax_count"
  pooling_param { 	pool: MAX 		kernel_size: 2		stride: 2 }
}

# unpool5
layer {
  name: "unpool5"	type: "Unpooling"	bottom: "pool5"		top: "unpool5"
						bottom: "pool5_mask"
						bottom: "pool5_argmax_count"
unpooling_param {   	unpool: MAX   		kernel_size: 2    	stride: 2 } #				unpool_size: 14 }
}`

**Comment by ChiZhangRIT** [comment]
This problem was solved on my machine. The pooling layer should have 3 top layers and corresponding unpooling layers have 3 bottoms as well, as Jeff said. Also, please note that in unpooling layers we use `pooling_param {...} `instead of `unpooling_param {...}` to determine the kernel size and stride.

**Comment by PeterJackNaylor** [comment]
Hi,

I have been using this unpooling layer, however I don't understand why it has been removed from master? Any reason?

**Comment by Timo-hab** [comment]
I have the same question. I can not find unpooling layer in master branch. Does it come back?

**Comment by VladTur** [comment]
Hi guys, sorry I did not respond you.
I have beed used a HyeonwooNoh's implementation for my CAE.

See details in
https://arxiv.org/abs/1701.04949

I have uploaded the models described in this paper to the Caffe User Group
https://groups.google.com/forum/#!topic/caffe-users/GhrCtONcRxY

Cheers,
Vlad

**Comment by belgraviton** [comment]
I check this patch with version 1. It's OK. It should be accepted to master.

**Comment by Timo-hab** [comment]
Is there a cuDNN version of unpooling layer as well?

**Comment by belgraviton** [comment]
Yes. In 'unpooling_layer.cu' source file.

**Comment by pauldelmusica** [comment]
I still don't see it in master. How can I use this feature? Is there a way to download these 8 files and merge them myself?

**Comment by VladTur** [comment]
Hi Paul,

No, these files are in Noh's branch, see
https://github.com/HyeonwooNoh/caffe/blob/master/src/caffe/layers/unpooling_layer.cpp

Yes, you can manually copy them to the <layers> folder on your machine and recompile your Caffe. I have tried this on different Caffe versions, everything works perfectly. 

Cheers,
Vlad

**Comment by ryanflower** [comment]
To apply the pull request, run the following command.
git clone https://github.com/BVLC/caffe.git # Clone the repository
git fetch origin pull/3601/head:unpooling # Fetch the pull #3601 to a local branch called "unpooling"
git cherry-pick 612227a666161ef1786cfdabdf4b055b23265445 # Cherry pick the commit to master

**Comment by pauldelmusica** [comment]
Thanks ryanflower,

The second command failed, as shown below.

c:\projects\caffe-unpool>git fetch origin pull/3601/head:unpooling
fatal: Not a git repository (or any of the parent directories): .git

**Comment by pauldelmusica** [comment]
Hi Vlad,

  Thanks for the pointer. What are the names of the files that are changed for this enhancement? I would prefer a simple copy->edit->compile sequence the old fashioned way. I don't plan to put anything back to Git.

  I understand the Noh code is a few years old. Would it be more difficult to merge it with the current "master"?

Thank you for your help.

**Comment by VladTur** [comment]
Hi Paul, Its 8 files in the enclosed .zip file. Short instruction in the link.txt.
Just copy them and recompile, I did it with the last version of Caffe several months ago, everything is worked. Vlad

[UnpoolingLayer.zip](https://github.com/BVLC/caffe/files/1029484/UnpoolingLayer.zip)

**Comment by pauldelmusica** [comment]
Hi Vlad,

   Thank you very much for sending the files for unpooling earlier today. They compiled and linked without any problems. 

   With these enhancements would your autoencoder work on the master branch? (https://groups.google.com/forum/#!topic/caffe-users/GhrCtONcRxY) I have downloaded your deepCAE5models.zip file, but they only contain prototext and scripts. I got the training data from http://data.pymvpa.org/datasets/mnist/ , in hdf5 format. Do you have trained models for the public? 

Thank you for sharing your work!

**Comment by VladTur** [comment]
Hi Paul,

Yes, these models are working on master branch and basically the results - how they work - are presented in our arXiv paper https://arxiv.org/abs/1701.04949
Yes, I did not include data in my deepCAE5models.zip file, because data are big.

I am not sure about hdf files from the link http://data.pymvpa.org/datasets/mnist/
What I did, I had MNIST in .mat format and I converted it to hdf5 myself.
If you want - I have those .mat MNIST files on my google drive, give me your e-mail address and I will give you an access and you can download them.

No, I did not publish trained models (.caffemodel and .solverstate) for the public, but if you need them - I can upload some of them on my google drive and give you an access.

Cheers,
Vlad

**Comment by pauldelmusica** [comment]
awesome! My email is paul.delamusica@gmail.com

Please include at least one trained model with the unpooling layer, along with your data file. The goal is to get the setup working. 

Thanks again!
paul.delamusica@gmail.com

**Comment by VladTur** [comment]
I saved one model - it corresponds to Model 4 and the result on Fig. 9 (right) in the paper.
There you can find original MNIST .mat files as well as my version of MNIST hdf5 files.
Cheers,
Vlad

**Comment by pauldelmusica** [comment]
Vlad,

Thanks.

Do I also need /global/home/vtu/caffemVTU/01data/trainnormfile.txt and
testnormfile.txt?

**Comment by VladTur** [comment]
No, it is just a path to a folder where h5 file is placed, actually this .txt file you feed to the Caffe model, it contains a list of all h5 files you want to use for the training or testing
like this
[trainnormfile-Hodgkin.txt](https://github.com/BVLC/caffe/files/1032769/trainnormfile-Hodgkin.txt)

**Comment by pauldelmusica** [comment]
Thank you, Vlad. 

In addition to the 8 files, what do I need to do in the caffe.proto file to tie-in the unpool layer? 

So far, I used Noh's caffe.proto as an example and did some modification such as adding UNPOOLING in the later type, a message for unpooling parameters, and an " optional UnpoolingParameter unpooling_param = 44;" However, I keep getting the following error and the program crashes:

error : Value of: V1LayerParameter_LayerType_IsValid(i)	test_upgrade_proto.cpp	2898 1 runtest

I am confused about LayerParameter, V1LayerParameter, and V0LayerParameter. It seems only V1LayerParameter has a LayerType definition and that is where the error message is emitting from.

**Comment by Teyen-Forde** [comment]
@VladTur @jeffdonahue Hello, 

I merged this commit and recompiled successfully even though there were some errors when I do "make test"(error with test_unpooling_layer.cpp). However, I am having some problem with setting up the layers. Caffe does not recognize "unpooling_param". I was wondering if you modified some other files, for example, caffe.proto.

Thanks

**Comment by belgraviton** [comment]
@Teyen-Forde You should use "pooling_param" for unpooling layer. This is an example:
`layer {
  name: "unpool4"
  type: "Unpooling"
  bottom: "conv5_1t"
  bottom: "conv4_3"
  top: "unpool4"
  pooling_param {
    pool: AVE
    kernel_size: 2
    stride: 2
  }
}`

**Comment by Teyen-Forde** [comment]
@belgraviton I went over some discussion about the usage of unpooling layer again. Many of them mentioned there are three bottom layers for unpooling: 1. normal inputs to be unpooled; 2. the pooling mask from previous pooling layer. 3. a shape of unpooling results.  I also noticed there is a unpooling parameter called unpool_size, which determines the output size, I guess. But when I included it, caffe does not recognize it again. How do you think about it?

I have tried your example. It came up with the same check error: Check failed: expected_pooled_height == pooled_height_ (8 vs. 16).

Are there some other things I am still missing?

Thanks!

**Comment by belgraviton** [comment]
@Teyen-Forde Max pooling use 3 bottom layers, average pooling - 2 bottom layers (inputs + shape).
Shapes of last bottom and top should be equal.

**Comment by Teyen-Forde** [comment]
yes, but for the non-overlapping case, we only need 2 bottom layers for Max (un)pooling and 1 bottom layer for AVE (un)pooling.

**Comment by belgraviton** [comment]
yes, last bottom is unnecessary. I use it in order to skip errors.

**Comment by VladTur** [comment]
To: @Teyen-Forde
Yes, @belgraviton is right, unpooling layer has NO <unpooling_param>, I do not know where you founded it. Similar to pooling, unpooling layer has <pooling_param>, see .prototxt files I have published, there are no <unpooling_param> there. Thanks Vadzim @belgraviton for answering this question! Cheers, Vlad

**Comment by pauldelmusica** [comment]
Thank you Vlad.

I added unpooling_param by combining the caffe.proto files of the current
release with that of Noh. I thought the 8 files you shared would work with
unpooling_param, since they are also in the input file deepCAE5models.zip.

So the task is in merging these 8 file with the master branch, along with
proper parameter names as in Noh's code.

**Comment by VladTur** [comment]
Paul, as I expained above I did not use "unpool_param" when I tested CAE on new version of Caffe with "layer" syntax. So I would like to check this again because I do not remember now <why?>. I will come back with this probably tomorrow or on Tuesday.

**Comment by belgraviton** [comment]
Is there python support for this unpooling layer?

**Comment by VladTur** [comment]
Hi Vadzim,
I think 'yes' since all of this working inside the Caffe version distributed by Noh.
I did all the visualizations in Matlab, but I think it can be done from python too.
Vlad

**Comment by belgraviton** [comment]
It's work fine. I had problems with this layer in python due to parameters. They should be set using special dict:
`n.pool2, n.pool2_m, n.pool2_argm = L.Pooling(n.conv2, kernel_size=2, pool=P.Pooling.MAX,ntop=3)`
`n.conv3 = L.Convolution(n.pool2, kernel_size=3, num_output=50)`
`n.unpool3 = L.Unpooling(n.conv3, n.pool2_m, n.pool2_argm, pooling_param=dict(kernel_size=2, pool=P.Pooling.MAX))`

**Comment by 07freedom** [comment]
Hi @VladTur,
I have merged the 8 files but run into error when compiling with "make all -j4".
It says "make: *** [.build_release/src/caffe/layers/unpooling_layer.o] Error 1"
Could you help me figure it out?

**Comment by Teyen-Forde** [comment]
@07freedom Have you checked out the zip file Vlad provided? There is a txt file inside telling where to put the 8 files. However, I think there is a minor error. Maybe you should add "test_unpooling_layer.cpp"  to "../src/caffe/test" folder.

**Comment by VladTur** [comment]
Yes, please check if you put those 9 files (see UnpoolingLayer.zip above) into three foldels how it was explained in the link.txt (this file is inside UnpoolingLayer.zip). Have you done this correctly?

**Comment by VladTur** [comment]
Sorry, 8 files, @07freedom is right!

**Comment by 07freedom** [comment]
@VladTur I believe the problem is protoc version mismatch. I tried 3.2 and 2.6 but the problem is still there. Could you tell me what is your protoc version?

**Comment by VladTur** [comment]
Hi @07freedom, I have libprotoc 2.6.1 on my current machine. I temporarily lost an access to my old machine, but I just have tested on my current machine with libprotoc 2.6.1 and Noh's Caffe version, everything is working perfectly. Cheers, Vlad

**Comment by bharatsau** [comment]
Hi @pauldelmusica,
I am having difficulty to merge two caffe.proto files(one is hnoh's, another is from latest caffe-master). 
Before trying for merging, I tried the other methods mentioned by the users in order to use unpooling layer, but all were unsuccessful.
I think the only way to progress is merging these two caffe.proto files. Please help me if you have found the solution. It will be great if you can share the merged caffe.proto file if you have.

Thanks

**Comment by 07freedom** [comment]
@Teyen-Forde 
I followed your instructions but it failed in "make runtest" and I have no idea what goes wrong. Any ideas?
F0816 15:44:09.233674 27197 unpooling_layer.cpp:45] Check failed: expected_pooled_height == pooled_height_ (0 vs. 6) 
*** Check failure stack trace: ***
    @     0x7f6dbd72d5cd  google::LogMessage::Fail()
    @     0x7f6dbd72f433  google::LogMessage::SendToLog()
    @     0x7f6dbd72d15b  google::LogMessage::Flush()
    @     0x7f6dbd72fe1e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f6dbaee7fa7  caffe::UnpoolingLayer<>::Reshape()
    @           0x47ecbb  caffe::Layer<>::SetUp()
    @           0x4f1d01  caffe::UnpoolingLayerTest<>::PoolUnpoolLayer::LayerSetUp()
    @           0x47eb5f  caffe::Layer<>::SetUp()
    @           0x4b0cdf  caffe::GradientChecker<>::CheckGradientExhaustive()
    @           0x52afb1  caffe::UnpoolingLayerTest<>::PoolUnpoolGradientCheck()
    @           0x968853  testing::internal::HandleExceptionsInMethodIfSupported<>()
    @           0x961e6a  testing::Test::Run()
    @           0x961fb8  testing::TestInfo::Run()
    @           0x962095  testing::TestCase::Run()
    @           0x96336f  testing::internal::UnitTestImpl::RunAllTests()
    @           0x963693  testing::UnitTest::Run()
    @           0x46f63d  main
    @     0x7f6dba171830  __libc_start_main
    @           0x4772d9  _start
    @              (nil)  (unknown)
Makefile:532: recipe for target 'runtest' failed
make: *** [runtest] Aborted (core dumped)

**Comment by 07freedom** [comment]
@VladTur 
And could you help me double check the instruction of Unpooling.zip?
Because there is only "util" not "utils" under "../include/caffe/" AND I am not so sure where I should put "test_unpooling_layer.cpp"
Thanks

**Comment by Teyen-Forde** [comment]
@07freedom 
As I mentioned before, you should add "test_unpooling_layer.cpp" to "../src/caffe/test" folder. And you are right, there is only "util" folder. That is where you put "*.cuh" files.

**Comment by 07freedom** [comment]
@Teyen-Forde 
Thanks for the quick reply. I can compile it but it failed to pass the test. I don't understand why the maxtops is still 2 why I have already changed the code.

**Comment by 07freedom** [comment]
Solved! Some rookie mistake in make/remake... Thanks guys

**Comment by belgraviton** [comment]
I'm include this PR to my Caffe v1.0 [fork](https://github.com/belgraviton/caffe).

**Comment by nmerrill67** [comment]
Why is this not merged?
